02-intermediate-ruby/api-auth.md

```diff
@@ -1,3 +1,226 @@
+
+### Example: The Movie DB
+
+[The Movie DB](https://www.themoviedb.org/) is an online resource for all things cinema. It's similar to IMDB, except IMDB is just a website, while The Movie DB has an API. That makes TMDB much more useful to us as programmers.
+
+Enter the following URL into your browser:
+
+```
+https://api.themoviedb.org/3/search/movie?query=hidden%20figures
+```
+
+**Question:** Did it work? What is the API telling you?
+
+The Movie DB requires an authentication token (what they call an "API key") before it will respond to requests. So let's go get one!
+
+1. [Create an account](https://www.themoviedb.org/account/signup)
+1. Find your API keys:
+    - Click on the picture of you in the upper right-hand corner
+    - In the drop-down menu, click `Settings`
+    - On the left, click `API`
+    - Your key is under `API Key (v3 auth)`. Copy it to your clipboard now.
+
+Now we can update the URL we used before with an extra `api_key` query parameter:
+
+```
+https://api.themoviedb.org/3/search/movie?query=hidden%20figures&api_key=<paste here>
+```
+
+By the way, all of this is covered in [The Movie DB's documentation](https://developers.themoviedb.org/3). The documentation is where you would go to figure all this out for a new API.
+
+### Resetting Keys
+
+Every once in a while you will find that your keys have been compromised and are no longer secret. Maybe you uploaded them to Git accidentally, they were in the background of a screenshot, or maybe you had them in your clipboard and accidentally pasted them into Slack. Now what?
+
+Fortunately most APIs make it easy to _reset_ your API keys. This will invalidate the old key and generate a new one for you to use. This is true of most of the APIs we'll be using in this class, including Slack and GitHub, but not TMDB.
+
+## Storing Credentials Securely
+
+This authorization token uniquely identifies and provides access to your account. Just like a password, anyone who has the token can pretend to be you. This means you need to be careful about keeping track of your tokens.
+
+In particular, you cannot commit tokens to git, because everything in a git repository is visible publicly on the internet. Instead we will make these tokens part of our application's _environment_.
+
+### Environment Variables
+
+Every program you run has access to a special set of information known as its environment. Some of this information is provided by the operating system, some by the Ruby interpreter, and some comes from the application itself. In Ruby, we can access this information through a built-in hash constant called `ENV`. Fire up `pry` and take a look at this constant:
```

Comment:
Nice, I didn't realize `ENV` had guts like this! This is a great intro to the topic